### PR TITLE
bpo-43406: Fix possible race condition where ``PyErr_CheckSignals`` tries to execute a non-Python signal handler

### PR DESCRIPTION
**Pull request by pitrou**
https://bugs.python.org/issue43406

### CODE REVIEW

File: Modules/signalmodule.c [hunk 1]
```diff
@@ -1706,10 +1706,27 @@ _PyErr_CheckSignalsTstate(PyThreadState *tstate)
...
+        PyObject *func = Handlers[i].func;
+        if (func == NULL || func == Py_None || func == IgnoreHandler ||
+            func == DefaultHandler) {
+            /* No Python signal handler, continue
+             * XXX this means the signal isn't handled by anyone
+             * (should we call raise() here?), but it's better than
+             * raising cryptic exceptions asynchronously
+             * such as "TypeError: 'int' object is not callable".
+             */
+            continue;
```
Comment (pablogsal) [inline Modules/signalmodule.c:1730]: This should probably require some consensus, but I would personally raise, as the situation is tricky enough that I think it should not pass silently.

Comment (pitrou) [reply Modules/signalmodule.c:1730]: @vstinner What do you think?

**Review by pablogsal — APPROVED** [review]
LGTM modulo a couple of comments

File: Lib/test/test_signal.py [hunk 2]
```diff
@@ -1251,6 +1252,44 @@ def handler(signum, frame):
         # Python handler
         self.assertEqual(len(sigs), N, "Some signals were lost")
 
+    def test_stress_modifying_handlers(self):
+        # bpo-43406: race condition between trip_signal() and signal.signal
+        signum = signal.SIGUSR1
```
Comment (eryksun) [inline Lib/test/test_signal.py:1259]: Skip this test if there's no `signal.SIGUSR1` (i.e. Windows).

File: Misc/NEWS.d/next/Core and Builtins/2021-03-04-22-53-10.bpo-43406.Na_VpA.rst [hunk 3]
```diff
@@ -0,0 +1,2 @@
+Fix possible race condition where ``PyErr_CheckSignals`` tries to execute a
+non-Python signal handler.
```
Comment (pablogsal) [inline Misc/NEWS.d/next/Core and Builtins/2021-03-04-22-53-10.bpo-43406.Na_VpA.rst]:
```suggestion
Fix a possible race condition where ``PyErr_CheckSignals`` tries to execute a
non-Python signal handler.
```

File: Modules/signalmodule.c (hunk 1 above)
Comment (pitrou) [reply Modules/signalmodule.c:1730]: One problem is that re-raising will break the assumption that `_thread.interrupt_main` only _simulates_ SIGINT.

Comment (pablogsal) [reply Modules/signalmodule.c:1730]: > One problem is that re-raising will break the assumption that `_thread.interrupt_main` only _simulates_ SIGINT.

That's an excellent point actually. Maybe we should set an unraisable exception (`PyErr_WriteUnraisable`)?

File: Lib/test/test_signal.py (hunk 2 above)
Comment (pitrou) [reply Lib/test/test_signal.py:1259]: Will do, thanks.

File: Modules/signalmodule.c (hunk 1 above)
Comment (pitrou) [reply Modules/signalmodule.c:1730]: That would not be much better than the asynchronous `TypeError`, would it?

Comment (pablogsal) [reply Modules/signalmodule.c:1730]: It does not set the error indicator IIRC and can be handled separately if needed by a hook. By default is like printing to stderr. The advantage would be that keeps the assumption that `_thread.interrupt_main` only simulates SIGINT but doesn't pass silently.

Comment (pitrou) [reply Modules/signalmodule.c:1730]: Ah, sounds good then!

Comment (pitrou) [reply Modules/signalmodule.c:1730]: Feel free to take a look at the updated patch @pablogsal .

**Review by pablogsal — APPROVED** [review]
LGTM

This is great work. Thanks a lot for working on this!

I think we can refine the error message (so we don't say "race condition" as users won't be able to act on that) but I propose to merge this as is and I will propose another PR with an improvement in the error message.

**Comment by pitrou** [comment]
Thanks for the quick review @pablogsal !

**Pull request merged**